### PR TITLE
fix(bug): chunkFilename as function doesn't work

### LINKED ISSUE #6639: chunkFilename - function doesn't work - Webpack 4
**Do you want to request a *feature* or report a *bug*?**
no

**What is the current behavior?**
Error when function is used as chunkFilename option.

**If the current behavior is a bug, please provide the steps to reproduce.**
* Create project that should have some chunks (something like `System.import('./test.js').then(function module() { console.log('OK'); });` )
* Set chunkFilename option to: `function(){ return '[id].chunk.js'; }`
* Try to compile.

**What is the expected behavior?**
Chunk with name "{some_id}.chunk.js" should be generated.

**If this is a feature request, what is motivation or use case for changing the behavior?**
no

**Please mention other relevant information such as the browser version, Node.js version, webpack version, and Operating System.**
The problem is that chunkFilename is handled incorrectly in "JsonpMainTemplatePlugin.js" and all other templates (and maybe somewhere else to).
For example here:
https://github.com/webpack/webpack/blob/5044762c846a3a83a208853e48785a293fe9ef7d/lib/web/JsonpMainTemplatePlugin.js#L63

There should be something like:

```
const chunkFilename = typeof path === "function" ? mainTemplate.outputOptions.chunkFilename(....required arguments....) : mainTemplate.outputOptions.chunkFilename; 
```

### PR DESCRIPTION
**Pull request by EugeneHlushko**
**What kind of change does this PR introduce?**
Fixes #6639 

**Did you add tests for your changes?**
no

**If relevant, link to documentation update:**

**Summary**
For technically reasons chunkFilename can't be a function. This PR disabling it on the the schema level.

**Does this PR introduce a breaking change?**
no

**Other information**

### CODE REVIEW

**Comment by EugeneHlushko** [comment]
Congrats on PR privileges 🍾 @ooflorent

**Comment by EugeneHlushko** [comment]
Hmm not sure i follow, if it throws here:
https://github.com/webpack/webpack/blob/edbb6f64bd630b89e9d228df0bec146efb40373c/lib/WebpackOptionsDefaulter.js#L88

means tests will fail for providing output.filename as a function in /tests/configCases/output/function

@sokra can you clarify?

**Comment by sokra** [comment]
@EugeneHlushko ok you are right. We can just stop defaulting it to a function.

You can skip the following lines when `filename` is a function:

https://github.com/webpack/webpack/blob/edbb6f64bd630b89e9d228df0bec146efb40373c/lib/WebpackOptionsDefaulter.js#L89-L93

**Comment by EugeneHlushko** [comment]
So i had to wrap 1 more line into the if statement
```
filename.replace(/(^|\/)([^/]*(?:\?|$))/, "$1[id].$2");
```

File: lib/WebpackOptionsDefaulter.js [hunk 1]
```diff
@@ -85,14 +85,15 @@ class WebpackOptionsDefaulter extends OptionsDefaulter {
...
-			return filename.replace(/(^|\/)([^/]*(?:\?|$))/, "$1[id].$2");
+			if (typeof filename !== "function") {
+				const hasName = filename.includes("[name]");
+				const hasId = filename.includes("[id]");
+				const hasChunkHash = filename.includes("[chunkhash]");
+				// Anything changing depending on chunk is fine
+				if (hasChunkHash || hasName || hasId) return filename;
+				// Elsewise prefix "[id]." in front of the basename to make it changing
+				return filename.replace(/(^|\/)([^/]*(?:\?|$))/, "$1[id].$2");
+			}
```
Comment (sokra) [inline lib/WebpackOptionsDefaulter.js]: You need to return something in the case of filename being a function too.

Maybe just return `[id].js`

**Comment by EugeneHlushko** [comment]
yeap looks good now

**Comment by sokra** [comment]
Thanks

**Pull request merged**